### PR TITLE
feat(core): add runReport to onTestRunEnd payload

### PR DESCRIPTION
**Pull request by fi3ework**
## Summary

Adds a precomputed `runReport` field to the `Reporter.onTestRunEnd` payload, so reporters and the CLI exit-code logic share one run-level summary instead of each filtering the raw `results` / `testResults` arrays. Matches the contract style used by Jest (`AggregatedResult`) and Vitest (`reason` + `unhandledErrors`).

## What this gives custom reporters

- New `runReport: RunReport` field carrying `status` (`'pass' | 'fail'`), pre-aggregated `counts`, the flattened `failures` list, the serialized `unhandledErrors`, and `duration` / `snapshot`.
- New public type exports from `@rstest/core`: `RunReport`, `FailureItem`, `FormattedError`.
- All previously-documented `onTestRunEnd` payload fields (`results`, `testResults`, `duration`, `snapshotSummary`, `unhandledErrors`, `filterRerunTestPaths`) are preserved — custom reporters that destructure only those keep compiling and behaving the same.

## What bug this prevents

Before this PR each reporter computed "did this run pass?" itself, and they had drifted: the markdown reporter ignored `passWithNoTests` in the no-tests case while the JSON reporter honored it, so the same run could report `pass` and `fail` in different outputs. With `buildRunReport` as the single source of truth this class of drift is no longer possible — adding a new reporter (or a new no-tests / unhandled-error edge case) only needs to be expressed once.

## Bug fixes bundled with this refactor

- **Multi-project runs with the same test path no longer overwrite each other's outcome.** `context.reporterResults` dedupes by `testPath`, so when two projects ran the same file the second project's status silently overwrote the first's. The run-level pass/fail predicate now reads from the local per-run arrays so it reflects every project's result. See the comment at `packages/core/src/core/runTests.ts` around the new `buildRunReport` call.
- **JUnit root `<testsuites>` counters now always match the emitted `<testcase>` body.** They previously could diverge from the body during watch reruns or multi-project same-path runs because the root used accumulated reporter state while the body used the current rerun's data. JUnit intentionally does **not** consume `runReport.counts` and instead derives counters from the same `testResults` it emits — this is the one place `runReport.counts` is *not* the source of truth, and there is an inline comment explaining why so it doesn't get "fixed" later.

## Behavior changes a reviewer should know about

- `runReport: RunReport` is **required** on the payload type (matches Jest's `AggregatedResult` and Vitest's `reason`, which are also required). TS structural typing means existing `implements Reporter` classes that don't destructure it keep compiling; the only break is hand-typing `Reporter['onTestRunEnd']` and manually calling it without the field — an extremely uncommon pattern.
- `JsonReporter`'s constructor no longer takes `config` (the `passWithNoTests` decision now lives in `buildRunReport`). Programmatic users who do `new JsonReporter({ config, rootPath, options })` need to drop the `config` argument; declarative reporter config in `rstest.config.ts` is unaffected.
- Watch-mode + no-tests now correctly leaves `process.exitCode` alone (it would have been set to `1` by the new predicate otherwise, contradicting `reportNoTestFiles`'s long-standing watch behavior).
- `runReport.unhandledErrors` preserves the full `FormattedError` shape (`diff`/`expected`/`actual`/`fullStack`), so reporters that print assertion diffs for run-level errors keep working. The `BlobReporter` payload and `mergeReports` were updated to round-trip these fields end-to-end.

## Compatibility

- No existing built-in reporter changes output for any input it previously handled correctly.
- Custom reporters using `Reporter` typing: no changes needed.
- Programmatic `new JsonReporter(...)` callers: drop the `config` argument.
- New docs in `website/docs/{en,zh}/api/javascript-api/{reporter,types}.mdx`.

## Checklist

- [x] Tests updated (or not required).
- [x] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/core/runTests.ts [hunk 1]
```diff
@@ -737,9 +750,24 @@ export async function runTests(context: Rstest): Promise<void> {
...
+        ? currentEntries.map((e) => e.testPath)
+        : undefined;
+
+      const runReport = buildRunReport({
+        // `context.reporterResults` dedupes by `testPath`, so in multi-project
+        // runs the same test file in a second project would overwrite the
+        // first project's status. Use the local per-run arrays so the run
+        // outcome reflects every project's result.
+        results,
+        testResults,
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/core/runTests.ts:763]: nitpick

**Comment by fi3ework** [comment]
this is a sub-topic of stabilizing the version 1.0.0 API. i'm closing this PR for now, we may cherry-pick some of the code once the 1.0 stable API is finalized.
